GitUI/CommandsDialogs/SettingsDialog/ShellExtension/ShellExtensionManager.cs

```diff
@@ -0,0 +1,112 @@
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using GitCommands;
+using Microsoft.Win32;
+
+namespace GitUI.CommandsDialogs.SettingsDialog.ShellExtension
+{
+    public static class ShellExtensionManager
+    {
+        /// <summary>
+        ///  Checks if the shell extension files are found at the known location.
+        /// </summary>
+        /// <returns><see langword="true"/> the extension files are found at the known location; otherwise <see langword="false"/>.</returns>
+        public static bool CheckFilesFound()
+        {
+            string path32 = FindFileInBinFolders(CommonLogic.GitExtensionsShellEx32Name);
+            string path64 = FindFileInBinFolders(CommonLogic.GitExtensionsShellEx64Name);
+            return !(string.IsNullOrEmpty(path32) || (Environment.Is64BitOperatingSystem && string.IsNullOrEmpty(path64)));
+        }
+
+        /// <summary>
+        ///  Checks if the shell extension is properly registered in the system.
+        /// </summary>
+        /// <returns><see langword="true"/> if the extension is registered; otherwise <see langword="false"/>.</returns>
+        public static bool CheckIfRegistered()
+        {
+            return !string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.LocalMachine, @"Software\Microsoft\Windows\CurrentVersion\Shell Extensions\Approved", "{3C16B20A-BA16-4156-916F-0A375ECFFE24}")) &&
+                !string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.ClassesRoot, @"*\shellex\ContextMenuHandlers\GitExtensions2")) &&
+                !string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.ClassesRoot, @"Directory\shellex\ContextMenuHandlers\GitExtensions2")) &&
+                !string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.ClassesRoot, @"Directory\Background\shellex\ContextMenuHandlers\GitExtensions2"));
+        }
+
+        /// <summary>
+        /// Register shell extensions
+        /// </summary>
+        /// <exception cref="FileNotFoundException">If at least one necessary for registration file wasn't found</exception>
+        /// <exception cref="Win32Exception">If user canceled elevation dialog</exception>
+        public static void Register() => RunRegSvrForShellExtensionDlls("{0}");
+
+        /// <summary>
+        /// Unregister shell extensions
+        /// </summary>
+        /// <exception cref="FileNotFoundException">If at least one necessary for registration file wasn't found</exception>
+        /// <exception cref="Win32Exception">If user canceled elevation dialog</exception>
+        public static void Unregister() => RunRegSvrForShellExtensionDlls("/u {0}");
```

Comment:
These are _possible_ causes for these exceptions. `Win32Exception` could also be thrown for other reasons - as well as other exceptions.
So please do not assert these were the only possible exceptions, and catch any `Exception`.
Perhaps `Win32Exception.NativeErrorCode` can be used to ignore a canceled elevation more specific.

Reply:
It's also just refactoring. I just noted the assumptions from code to xmldoc.
As I don't want to change old behavior I don't think it is good idea to change this logic.

Reply:
P.S. Or did you mean I should encapsulate exception logic in ShellExtensionManager and return something like error code instead?

Reply:
> P.S. Or did you mean I should encapsulate exception logic in ShellExtensionManager and return something like error code instead?

No, I do not want error codes.

With this refactoring, you have created a public API,
and added xmldoc - good,
but this xmldoc promises that these functions would throw only the listed exceptions - wrong.

I did not review the original code and would not have accepted it.

Reply:
My primary intension is not to break working code.
1. I don't know whether author wanted to catch only elevation canceled exception, or any Win32Exception.
2. I'm not sure if for every OS and all possible UAC settings canceling elevation dialog will lead to the same exception.

So I don't understand what should I do. Change xml doc? Change logic? Anyway, please, suggest changes as you more familiar with the code.

Reply:
Please do not assert in xmldoc that there were no other possible exceptions. The list must be complete.
Swallowing errors is a bad practice. So catch and display all exceptions except for canceled user elevation.
Done in https://github.com/mstv/gitextensions/commit/61c09dc0b0c54fb02ce437523d1d3969be7cfc65

Reply:
I introduced ShellExtensionManager as "model" layer api. With your change it became "presentation" layer api (as it can show message boxes). I don't like the idea.
The worse part is that now the message box will be shown during PageToSettings.
So I see at least those issues:
1. When message is shown the settings can be in inconsistent state (as message box will appear at random time during saving settings). So the settings will be inconsistent as long as user don't close message box (if ever).
2. It more UX issue but still, when user changes page and clicks save he will see "irrelevant" error. I call it "irrelevant" as the user can do nothing with this error. The settings window is closed, so he will have to open it again and go to proper page to find what have happened.

That's what I meant, when I said that I don't want to break working code and wouldn't have agreed with the change if I had reviewed it.

As for me I'd suggested rolling back `ShowMessage` call from `ShellExtensionManager`, rewrite `FormSettings::Save` so that it will:
1. Catch exceptions from `settingsPage.SaveSettings();`
2. If any exception were caught, continue saving all other settings (or rollback already saved changes), so that settings will be in consistent state.
3. Show error and navigate to page with the issues without closing the setting dialog, so user can see the problem an fix it.

Reply:
P.S. Also, if you decided to show all errors to the user I believe you should show elevation cancelled error message to the user. If he rejected elevation he will get clear message why we cannot perform requested action.

Reply:
> I introduced ShellExtensionManager as "model" layer api. With your change it became "presentation" layer api (as it can show message boxes). I don't like the idea.

Yes, it was a quick draft. The exception handling could be moved to a dedicated class.
The settings page belongs to the "presentation" layer.

> The worse part is that now the message box will be shown during PageToSettings. So I see at least those issues:
> 
>     1. When message is shown the settings can be in inconsistent state (as message box will appear at random time during saving settings). So the settings will be inconsistent as long as user don't close message box (if ever).
>
>     2. It's more UX issue but still, when user changes page and clicks save he will see "irrelevant" error. I call it "irrelevant" as the user can do nothing with this error. The settings window is closed, so he will have to open it again and go to proper page to find what have happened.

With the current approach, the user must reopen the settings anyway because - due to the swallowed exceptions - they cannot know whether it has worked or not.
IMO, the root cause for the UX issues is: This de/registration of the Explorer integration is *not* an actual GE setting, i.e. it is not stored in the settings file.
The de/registration should be done at once - not when the settings are saved. One or two Buttons instead of the CheckBox will give a clear UX (including the possibility to show error messages).
Errors should be displayed by extending the class `MessageBoxes`.

> P.S. Also, if you decided to show all errors to the user I believe you should show elevation cancelled error message to the user too. If he rejected elevation he will get clear message why we cannot perform requested action.

OK for me.